autoblocks/_impl/tracer.py

```diff
@@ -101,38 +110,119 @@ def start_span(self):
+        else:
+            try:
+                ctx = contextvars.copy_context()
+                evaluation = await global_state.event_loop().run_in_executor(
+                    None,
+                    ctx.run,
+                    evaluator.evaluate_event,
+                    event,
+                )
+            except Exception as err:
+                log.error("Unable to execute evaluation. Error: %s", repr(err), exc_info=True)
+
+        return evaluation
+
+    async def _run_evaluators(
+        self, evaluators: List[BaseEventEvaluator], event: TracerEvent, max_evaluator_concurrency: int
+    ) -> List[Dict[str, Any]]:
+        if len(evaluators) == 0:
+            return []
+        try:
+            evaluations: List[Union[Evaluation, None]] = await gather_with_max_concurrency(
+                max_evaluator_concurrency,
+                [
+                    self._evaluate_event(
+                        event=event,
+                        evaluator=evaluator,
+                    )
+                    for evaluator in evaluators
+                ],
+            )
+            return [
+                self._evaluation_to_json(event_evaluation=item, evaluator_external_id=evaluator.id)
+                for evaluator, item in zip(evaluators, evaluations)
+                if item is not None
+            ]
+
+        except Exception as err:
+            log.error("Unable to complete evaluating events. Error: %s", str(err), exc_info=True)
+            return []
+
     async def _send_event_unsafe(
         self,
         # Require all arguments to be specified via key=value
         *,
         message: str,
-        trace_id: Optional[str] = None,
-        span_id: Optional[str] = None,
-        parent_span_id: Optional[str] = None,
-        timestamp: Optional[str] = None,
-        properties: Optional[Dict] = None,
-        prompt_tracking: Optional[Dict] = None,
+        trace_id: Optional[str],
+        span_id: Optional[str],
+        parent_span_id: Optional[str],
+        timestamp: Optional[str],
+        properties: Optional[Dict],
+        prompt_tracking: Optional[Dict],
+        evaluators: Optional[List[BaseEventEvaluator]],
+        max_evaluator_concurrency: int,
     ) -> SendEventResponse:
         merged_properties = dict(self._properties)
         merged_properties.update(properties or {})
+
+        # Overwrite properties with any provided as top level arguments
         if span_id:
             merged_properties["span_id"] = span_id
         if parent_span_id:
             merged_properties["parent_span_id"] = parent_span_id
         if prompt_tracking:
             merged_properties["promptTracking"] = prompt_tracking
-
         trace_id = trace_id or self._trace_id
         timestamp = timestamp or datetime.now(timezone.utc).isoformat()
 
+        # If there are evaluators, run them and compute the evaluations property
+        if evaluators:
+            evaluations = await self._run_evaluators(
```

Comment:
i think we actually need a try-except on this as well, since we're only catching failures in the user-defined `evaluate_event`. but if there's an error in our code it will prevent the event from being sent

for example, maybe add a test where our `_evaluation_to_json` method raises an error and assert that the event is still sent

Reply:
in python you can use mock.patch to patch the tracer's method to raise